sdk/core/azure-core/src/uuid.cpp

```diff
@@ -18,84 +23,158 @@ static thread_local std::mt19937_64 randomGenerator(std::random_device{}());
+            parseError = true;
+            break;
+          }
+
+          result.m_uuid[bi] = (HexToNibble(c) << 4) | HexToNibble(c2);
+          ++si;
+          ++bi;
+        }
+      }
+    }
+
+    return parseError ? throw std::invalid_argument(
+               "Error parsing Uuid: '" + s
+               + "' is not in the '00112233-4455-6677-8899-aAbBcCdDeEfF' format.")
+                      : result;
+  }
+
   Uuid Uuid::CreateUuid()
   {
-    uint8_t uuid[UuidSize] = {};
+    Uuid result{};
+
+    // Using RngResultType and RngResultSize to highlight the places where the same type is used.
+    using RngResultType = std::uint32_t;
+    static_assert(sizeof(RngResultType) == 4, "sizeof(RngResultType) must be 4.");
+    constexpr size_t RngResultSize = 4;
 
 #if defined(AZ_PLATFORM_WINDOWS)
     std::random_device rd;
+
+    static_assert(
+        std::is_same<RngResultType, decltype(rd())>::value,
+        "random_device::result_type must be of RngResultType.");
 #else
-    std::uniform_int_distribution<uint32_t> distribution;
+    std::uniform_int_distribution<RngResultType> distribution;
 #endif
 
-    for (size_t i = 0; i < UuidSize; i += 4)
+    for (size_t i = 0; i < result.m_uuid.size(); i += RngResultSize)
     {
 #if defined(AZ_PLATFORM_WINDOWS)
-      const uint32_t x = rd();
+      const RngResultType x = rd();
 #else
-      const uint32_t x = distribution(randomGenerator);
+      const RngResultType x = distribution(randomGenerator);
 #endif
-      std::memcpy(uuid + i, &x, 4);
+      std::memcpy(result.m_uuid.data() + i, &x, RngResultSize);
     }
 
     // The variant field consists of a variable number of the most significant bits of octet 8 of
     // the UUID.
-    // https://www.rfc-editor.org/rfc/rfc4122.html#section-4.1.1
-    // For setting the variant to conform to RFC4122, the high bits need to be of the form 10xx,
+    // https://www.rfc-editor.org/rfc/rfc9562.html#name-variant-field
+    // For setting the variant to conform to RFC9562, the high bits need to be of the form 10xx,
     // which means the hex value of the first 4 bits can only be either 8, 9, A|a, B|b. The 0-7
     // values are reserved for backward compatibility. The C|c, D|d values are reserved for
     // Microsoft, and the E|e, F|f values are reserved for future use.
     // Therefore, we have to zero out the two high bits, and then set the highest bit to 1.
-    uuid[8] = (uuid[8] & 0x3F) | 0x80;
+    result.m_uuid.data()[8] = (result.m_uuid.data()[8] & 0x3F) | 0x80;
 
-    constexpr uint8_t version = 4; // Version 4: Pseudo-random number
-
-    uuid[6] = (uuid[6] & 0xF) | (version << 4);
-
-    return Uuid(uuid);
-  }
+    {
+      // https://www.rfc-editor.org/rfc/rfc9562.html#name-version-field
+      constexpr std::uint8_t Version = 4; // Version 4: Pseudo-random number
+      result.m_uuid.data()[6] = (result.m_uuid.data()[6] & 0xF) | (Version << 4);
+    }
 
-  Uuid Uuid::CreateFromArray(std::array<uint8_t, UuidSize> const& uuid)
```

Comment:
Why move the impl to the header file? Trying to understand the benefit of either approach. Do we prefer implementation to be in cpp file over hpp file or vice versa?

Reply:
Having it in `.hpp` file allows us to make it `constexpr`, because `std::array` initialization from another `std::array` is a constexpr, and is trivial in terms of assembly commands. The implementation is not multiple lines long, so should it be just inlined, the code size impact is tiny, probably the same number of assembler commands as function call, if not less. And also it does not require any extra includes to be made in the `.hpp` file, so compilation cost wise it is pretty lightweight as well.

I know that we are sending stuff over the network, so hardly any place would ever be a bottleneck in our SDK, yet I think it makes sense to not pessimize implementation, especially if it comes basically free and we can use it easily.
I mean, the other day, you would make a PR optimizing the `Uuid::ToString()` performance, and I and the team liked the change and agreed to it (caching was another thing, chacing would have tradeoffs, while writing simple parser instead of `sscanf()` did not have tradeoffs).